sensors/mag/lsm9dsxx.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Driver for IMU lsm9dsxx
+ * 
+ * lsm9dsxx is preset to:
+ *  - output data rate (ODR) to 952Hz 
+ *  - accelerometer range to +-8G
+ *  - gyroscope range to +-2000dps
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Hubert Buczynski, Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/time.h>
+#include <sys/threads.h>
+#include <spi.h>
+#include <spi-msg.h>
+#include <string.h>
+
+#include "../sensors.h"
+
+/* self-identification register of magnetometer */
+#define REG_WHOAMI     0x0f
+#define REG_VAL_WHOAMI 0x3D
+
+/* control register 1 */
+#define REG_CTRL_REG1_M                  0x20
+#define VAL_CTRL_REG1_M_XYHIGH_PERF_MODE 0x60
+#define VAL_CTRL_REG1_M_ODR_80           0x1c
+#define VAL_CTRL_REG1_M_FAST_ODR         0x02
+#define VAL_CTRL_REG1_M_SELF_TEST        0x01
+
+/* control register 2 */
+#define REG_CTRL_REG2_M       0x21
+#define MASK_CTRL_REG2_M_FS   0x60
+#define VAL_CTRL_REG2_M_FS_4  0x00
+#define VAL_CTRL_REG2_M_FS_8  0x20
+#define VAL_CTRL_REG2_M_FS_12 0x40
+#define VAL_CTRL_REG2_M_FS_16 0x60
+
+/* control register 3 */
+#define REG_CTRL_REG3_M               0x22
+#define MASK_CTRL_REG2_M_MD           0x03
+#define VAL_CTRL_REG3_M_MD_CONTINUOUS 0x00
+#define VAL_CTRL_REG3_M_MD_SINGLE     0x01
+#define VAL_CTRL_REG3_M_MD_PWR_DOWN   0x03
+#define VAL_CTRL_REG3_M_SIM_SPIRW     0x04
+#define VAL_CTRL_REG3_M_I2CDSBL       0x80
+
+/* control register 4 */
+#define REG_CTRL_REG4_M                 0x23
+#define VAL_CTRL_REG4_M_ZHIGH_PERF_MODE 0x08
+
+/* status registers */
+#define REG_STATUS_REG_M       0x27
+#define VAL_STATUS_REG_M_ZYXDA 0x08
+
+/* data storage addresses */
+#define REG_DATA_OUT         0x28
+#define SPI_READ_BIT         0x80
+#define SPI_AUTOADDRINCR_BIT 0x40
+
+/* sensors data sizes */
+#define SENSOR_OUTPUT_SIZE 6
+
+/* conversions */
+#define MAG4G_CONV_MGAUSS 0.14F        /* conversion from sensor (at +-4 gauss scale) value to milligauss */
+#define MGAUSS_2_UTSL     1.221730475F /* milligauss to microtesla convertion parameter */
+#define UTSL_2_MAGSHUB    10.F         /* microtesla to sensorhub magnetic flux storage scale of 10^-7 T */
+
+
+typedef struct {
+	spimsg_ctx_t spiCtx;
+	sensor_event_t evt;
+	char stack[512] __attribute__((aligned(8)));
+} lsm9dsxxMag_ctx_t;
+
+
+static int16_t translateMag(uint8_t hbyte, uint8_t lbyte)
+{
+	int16_t val = 0;
+	val |= hbyte;
+	val <<= 8;
+	val |= lbyte;
+	return val * MAG4G_CONV_MGAUSS * MGAUSS_2_UTSL * UTSL_2_MAGSHUB; /* value -> milligauss -> microtesla -> sensorhub storage scale*/
+}
+
+
+static int spiWriteReg(spimsg_ctx_t *spiCtx, uint8_t regAddr, uint8_t regVal)
+{
+	unsigned char cmd[2] = { (regAddr & 0x7F), regVal }; /* write bit set to regAddr */
```

Comment:
I don't understand this comment. Is it necessary?

Reply:
Changed comment to: `/* set first transmitted bit of regAddr to 0 to indicate write operation */`. 

I thought that some random bitewise operation on register address might need explanation. Without it we might randomly read something into not existing buffer passed to xfer.

Reply:
Removed all together after consulations